File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.cc
#include<iostream>

using namespace std;

int arr[501][501] = { 0, };
int dp[501][501] = { 0, };

int main() {

	int n;
	
	int sol = 0;

	cin >> n;

	for (int i = 1; i <= n; i++) {
		for (int j = 1; j <= i; j++) {
			cin >> arr[i][j];
		}
	}

	dp[1][1] = arr[1][1];

	for (int i = 2; i <= n; i++) {
		for (int j = 1; j < i; j++) {
			if (dp[i][j] < dp[i - 1][j] + arr[i][j]) {
				dp[i][j] = dp[i - 1][j] + arr[i][j];
			}
			if (dp[i][j+1] < dp[i - 1][j] + arr[i][j+1]) {
				dp[i][j+1] = dp[i - 1][j] + arr[i][j+1];
			}
		}
	}

	for (int i = 1; i <= n; i++) {
		if (sol < dp[n][i]) {
			sol = dp[n][i];
		}
	}

	cout << sol;

	return 0;
}

File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.java
import java.io.*;

class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int n = Integer.parseInt(br.readLine());

        int[][] arr = new int[n+1][n+2];
        int[][] dp = new int[n+1][n+2];

        int sol=0;

        // arr 채우기
        for (int i = 1; i <= n; i++) {
            String[] str = (br.readLine()).split(" ");
            for (int j = 1; j <= i; j++) {
                arr[i][j] = Integer.parseInt(str[j-1]);
            }
        }

        if (n == 1) {
            sol = arr[1][1];
        }
        else {
            for (int i = 1; i <= n; i++) {
                dp[n][i] = arr[n][i];
            }

            for (int i = n-1; i >= 1; i--) {
                for (int j = 1; j <= i; j++) {
                    dp[i][j] = Math.max(arr[i][j] + dp[i+1][j], arr[i][j] + dp[i+1][j+1]);
                }
            }
            sol = dp[1][1];
        }

        bw.write(sol+"\n");
        bw.flush();
        bw.close();

    }
}